File: Other Versions/Go/scanners/openclaw.go
package scanners

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

// secretKeys are the JSON keys that indicate a credential value.
var openclawSecretKeys = []string{
	"accessToken", "access_token", "refreshToken", "refresh_token",
	"apiKey", "api_key", "token", "auth_token", "secret",
	"password", "botToken", "bot_token", "clientSecret", "client_secret",
}

// secretKeywords used for heuristic key-name matching.
var openclawSecretKeywords = []string{
	"token", "key", "secret", "password", "passwd", "auth",
	"credential", "cred", "apikey", "api_key", "accesstoken",
	"refreshtoken", "bottoken", "clientsecret",
}

type openclawScanner struct{}

func init() {
	Register(&openclawScanner{})
}

func (s *openclawScanner) Name() string      { return "OpenClaw" }
func (s *openclawScanner) Slug() string       { return "openclaw" }
func (s *openclawScanner) IsApplicable() bool { return true }

func (s *openclawScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, base := range s.getBasePaths(plat) {
		s.scanAuthProfiles(base, &result, showSecrets)
		s.scanCredentialsDir(base, &result, showSecrets)
		s.scanSecretsJSON(base, &result, showSecrets)
		s.scanMainConfig(base, &result, showSecrets)
		s.scanEnvFile(base, &result, showSecrets)
		s.scanLegacyOauth(base, &result, showSecrets)
	}

	return result
}

func (s *openclawScanner) getBasePaths(plat core.Platform) []string {
	home := core.GetHome()
	paths := []string{filepath.Join(home, ".openclaw")}

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".openclaw"))
		}
	}

	return paths
}

// --- Auth Profiles (per-agent OAuth + API keys) ---

func (s *openclawScanner) scanAuthProfiles(base string, result *core.ScanResult, showSecrets bool) {
	agentsDir := filepath.Join(base, "agents")
	entries, err := os.ReadDir(agentsDir)
	if err != nil {
		return
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		authFile := filepath.Join(agentsDir, entry.Name(), "agent", "auth-profiles.json")
		data, err := os.ReadFile(authFile)
		if err != nil {
			continue
		}

		perms := core.GetFilePermissions(authFile)
		owner := core.GetFileOwner(authFile)

		var obj map[string]interface{}
		if err := json.Unmarshal(data, &obj); err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("Failed to parse %s: %s", authFile, err))
			continue
		}

		s.extractSecretsRecursive(obj, authFile, perms, owner, result, showSecrets, fmt.Sprintf("agent:%s", entry.Name()), 0)
	}
}

// --- Credentials directory ---

func (s *openclawScanner) scanCredentialsDir(base string, result *core.ScanResult, showSecrets bool) {
	credsDir := filepath.Join(base, "credentials")
	if _, err := os.Stat(credsDir); err != nil {
		return
	}

	// WhatsApp credentials
	waDir := filepath.Join(credsDir, "whatsapp")
	if entries, err := os.ReadDir(waDir); err == nil {
		for _, entry := range entries {
			if !entry.IsDir() {
				continue
			}
			credsFile := filepath.Join(waDir, entry.Name(), "creds.json")
			s.scanJSONFile(credsFile, result, showSecrets, fmt.Sprintf("whatsapp:%s", entry.Name()))
		}
	}

	// All JSON files in credentials root
	matches, _ := filepath.Glob(filepath.Join(credsDir, "*.json"))
	for _, jsonFile := range matches {
		s.scanJSONFile(jsonFile, result, showSecrets, "credentials")
	}
}

// --- secrets.json ---

func (s *openclawScanner) scanSecretsJSON(base string, result *core.ScanResult, showSecrets bool) {
	path := filepath.Join(base, "secrets.json")
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)

	var obj map[string]interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Failed to parse %s: %s", path, err))
		return
	}

	s.extractSecretsRecursive(obj, path, perms, owner, result, showSecrets, "secrets.json", 0)
}

// --- Main config (openclaw.json) ---

func (s *openclawScanner) scanMainConfig(base string, result *core.ScanResult, showSecrets bool) {
	path := filepath.Join(base, "openclaw.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)

	// Strip JS-style // comments
	var lines []string
	for _, line := range strings.Split(string(raw), "\n") {
		stripped := strings.TrimLeft(line, " \t")
		if strings.HasPrefix(stripped, "//") {
			continue
		}
		lines = append(lines, line)
	}

	var obj map[string]interface{}
	if err := json.Unmarshal([]byte(strings.Join(lines, "\n")), &obj); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Failed to parse %s: %s", path, err))
		return
	}

	// Check gateway auth token
	if gateway, ok := obj["gateway"].(map[string]interface{}); ok {
		if auth, ok := gateway["auth"].(map[string]interface{}); ok {
			if token, ok := auth["token"].(string); ok && len(token) > 8 {
				rawValue := ""
				if showSecrets {
					rawValue = token
				}
				notes := []string{"OpenClaw gateway auth token (inline)"}
				if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
					notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
				}
				result.Findings = append(result.Findings, core.CredentialFinding{
					ToolName:        s.Name(),
					CredentialType:  "gateway_auth_token",
					StorageType:     core.PlaintextJSON,
					Location:        path,
					Exists:          true,
					RiskLevel:       core.AssessRisk(core.PlaintextJSON, path),
					ValuePreview:    core.MaskValue(token, showSecrets),
					RawValue:        rawValue,
					FilePermissions: perms,
					FileOwner:       owner,
					FileModified:    core.GetFileMtime(path),
					Remediation:     "Use SecretRef (env:, file:) instead of inline secrets",
					RemediationHint: remediation.HintManual(
						"Use SecretRef (env:, file:) instead of inline secrets",
						map[string]any{"suggested_format": "env:VAR_NAME"},
					),
					Notes:           notes,
				})
			}
		}
	}

	// Check channel configs for inline tokens
	if channels, ok := obj["channels"].(map[string]interface{}); ok {
		s.extractSecretsRecursive(channels, path, perms, owner, result, showSecrets, "channels", 0)
	}

	// Check agent model configs for API keys
	if agents, ok := obj["agents"].(map[string]interface{}); ok {
		s.extractSecretsRecursive(agents, path, perms, owner, result, showSecrets, "agents", 0)
	}
}

// --- .env file ---

func (s *openclawScanner) scanEnvFile(base string, result *core.ScanResult, showSecrets bool) {
	path := filepath.Join(base, ".env")
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	storage := core.PlaintextENV

	secretPatterns := []string{"KEY", "TOKEN", "SECRET", "PASSWORD", "AUTH", "CREDENTIAL"}
	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if idx := strings.Index(line, "="); idx >= 0 {
			key := strings.TrimSpace(line[:idx])
			value := strings.TrimSpace(line[idx+1:])
			value = strings.Trim(value, "'\"")
			if value == "" {
				continue
			}
			keyUpper := strings.ToUpper(key)
			matched := false
			for _, p := range secretPatterns {
				if strings.Contains(keyUpper, p) {
					matched = true
					break
				}
			}
			if matched {
				rawValue := ""
				if showSecrets {
					rawValue = value
				}
				notes := []string{"From OpenClaw .env file"}
				if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
					notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
				}
				result.Findings = append(result.Findings, core.CredentialFinding{
					ToolName:        s.Name(),
					CredentialType:  fmt.Sprintf("env_file:%s", key),
					StorageType:     storage,
					Location:        path,
					Exists:          true,
					RiskLevel:       core.AssessRisk(storage, path),
					ValuePreview:    core.MaskValue(value, showSecrets),
					RawValue:        rawValue,
					FilePermissions: perms,
					FileOwner:       owner,
					FileModified:    core.GetFileMtime(path),
					Remediation:     "Use SecretRef (env:, file:) instead of inline secrets",
					RemediationHint: remediation.HintManual(
						"Use SecretRef (env:, file:) instead of inline secrets",
						map[string]any{"suggested_format": "env:VAR_NAME"},
					),
					Notes:           notes,
				})
			}
		}
	}
}

// --- Legacy oauth.json ---

func (s *openclawScanner) scanLegacyOauth(base string, result *core.ScanResult, showSecrets bool) {
	path := filepath.Join(base, "credentials", "oauth.json")
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)

	var obj map[string]interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		return
	}

	s.extractSecretsRecursive(obj, path, perms, owner, result, showSecrets, "legacy_oauth", 0)
}

// --- Helpers ---

func (s *openclawScanner) scanJSONFile(path string, result *core.ScanResult, showSecrets bool, context string) {
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)

	var obj map[string]interface{}
	if err := json.Unmarshal(data, &obj); err != nil {
		return
	}

	s.extractSecretsRecursive(obj, path, perms, owner, result, showSecrets, context, 0)
}

func (s *openclawScanner) extractSecretsRecursive(
	data map[string]interface{}, path, perms, owner string,
	result *core.ScanResult, showSecrets bool,
	context string, depth int,
) {
	if depth > 10 {
		return
	}

	for key, val := range data {
		switch v := val.(type) {
		case string:
			if len(v) <= 8 {
				continue
			}
			if !s.isSecretKey(key) && !s.looksLikeToken(v) {
				continue
			}

			// SecretRef detection: skip values referencing external sources
			if strings.HasPrefix(v, "env:") || strings.HasPrefix(v, "file:") || strings.HasPrefix(v, "exec:") {
				var notes []string
				if context != "" {
					notes = append(notes, fmt.Sprintf("Context: %s", context))
				}
				notes = append(notes, "SecretRef (not inline — references external source)")
				if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
					notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
				}
				result.Findings = append(result.Findings, core.CredentialFinding{
					ToolName:       s.Name(),
					CredentialType: fmt.Sprintf("secret_ref:%s", key),
					StorageType:    core.PlaintextJSON,
					Location:       path,
					Exists:         true,
					RiskLevel:      core.RiskInfo,
					ValuePreview:   v,
					FileModified:   core.GetFileMtime(path),
					Notes:          notes,
				})
				continue
			}

			storage := core.PlaintextJSON
			var notes []string
			if context != "" {
				notes = append(notes, fmt.Sprintf("Context: %s", context))
			}
			notes = append(notes, "Plaintext credential in OpenClaw config")
			if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
				notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
			}

			rawValue := ""
			if showSecrets {
				rawValue = v
			}
			result.Findings = append(result.Findings, core.CredentialFinding{
				ToolName:        s.Name(),
				CredentialType:  key,
				StorageType:     storage,
				Location:        path,
				Exists:          true,
				RiskLevel:       core.AssessRisk(storage, path),
				ValuePreview:    core.MaskValue(v, showSecrets),
				RawValue:        rawValue,
				FilePermissions: perms,
				FileOwner:       owner,
				FileModified:    core.GetFileMtime(path),
				Remediation:     "Use SecretRef (env:, file:) instead of inline secrets",
				RemediationHint: remediation.HintManual(
					"Use SecretRef (env:, file:) instead of inline secrets",
					map[string]any{"suggested_format": "env:VAR_NAME"},
				),
				Notes:           notes,
			})

		case map[string]interface{}:
			s.extractSecretsRecursive(v, path, perms, owner, result, showSecrets, context, depth+1)

		case []interface{}:
			for _, item := range v {
				if m, ok := item.(map[string]interface{}); ok {
					s.extractSecretsRecursive(m, path, perms, owner, result, showSecrets, context, depth+1)
				}
			}
		}
	}
}

func (s *openclawScanner) isSecretKey(key string) bool {
	keyLower := strings.ToLower(key)
	for _, kw := range openclawSecretKeywords {
		if strings.Contains(keyLower, kw) {
			return true
		}
	}
	return false
}

func (s *openclawScanner) looksLikeToken(value string) bool {
	if len(value) < 20 {
		return false
	}
	for _, prefix := range []string{"/", "http://", "https://", "env:", "file:", "exec:"} {
		if strings.HasPrefix(value, prefix) {
			return false
		}
	}
	alphanumCount := 0
	for _, c := range value {
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' {
			alphanumCount++
		}
	}
	return float64(alphanumCount)/float64(len(value)) > 0.8
}

File: aihound/scanners/openclaw.py
"""Scanner for OpenClaw AI agent platform credentials."""

from __future__ import annotations

import json
import logging
from pathlib import Path

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType, RiskLevel,
)
from aihound.core.platform import (
    detect_platform, Platform, get_home, get_wsl_windows_home,
)
from aihound.core.redactor import mask_value
from aihound.core.permissions import get_file_permissions, get_file_owner, assess_risk, get_file_mtime, describe_staleness
from aihound.remediation import hint_manual
from aihound.scanners import register

logger = logging.getLogger("aihound.scanners.openclaw")

# Token/secret keys to look for in JSON config files
SECRET_KEYS = [
    "accessToken", "access_token", "refreshToken", "refresh_token",
    "apiKey", "api_key", "token", "auth_token", "secret",
    "password", "botToken", "bot_token", "clientSecret", "client_secret",
]


@register
class OpenClawScanner(BaseScanner):
    def name(self) -> str:
        return "OpenClaw"

    def slug(self) -> str:
        return "openclaw"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        for base in self._get_base_paths(plat):
            self._scan_auth_profiles(base, result, show_secrets)
            self._scan_credentials_dir(base, result, show_secrets)
            self._scan_secrets_json(base, result, show_secrets)
            self._scan_main_config(base, result, show_secrets)
            self._scan_env_file(base, result, show_secrets)
            self._scan_legacy_oauth(base, result, show_secrets)

        return result

    def _get_base_paths(self, plat: Platform) -> list[Path]:
        paths = [get_home() / ".openclaw"]

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".openclaw")

        return paths

    # --- Auth Profiles (per-agent OAuth + API keys) ---

    def _scan_auth_profiles(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        agents_dir = base / "agents"
        if not agents_dir.exists():
            logger.debug("OpenClaw agents dir not found: %s", agents_dir)
            return

        for agent_dir in agents_dir.iterdir():
            if not agent_dir.is_dir():
                continue
            auth_file = agent_dir / "agent" / "auth-profiles.json"
            if not auth_file.exists():
                continue

            logger.debug("Reading auth profiles: %s", auth_file)
            perms = get_file_permissions(auth_file)
            owner = get_file_owner(auth_file)
            mtime = get_file_mtime(auth_file)

            try:
                data = json.loads(auth_file.read_text(encoding="utf-8"))
            except (json.JSONDecodeError, OSError) as e:
                logger.warning("Failed to parse %s: %s", auth_file, e, exc_info=True)
                result.errors.append(f"Failed to parse {auth_file}: {e}")
                continue

            agent_name = agent_dir.name
            self._extract_secrets_recursive(
                data, auth_file, perms, owner, mtime, result, show_secrets,
                context=f"agent:{agent_name}",
            )

    # --- Credentials directory (WhatsApp, Telegram, channel allowlists) ---

    def _scan_credentials_dir(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        creds_dir = base / "credentials"
        if not creds_dir.exists():
            logger.debug("OpenClaw credentials dir not found: %s", creds_dir)
            return

        logger.debug("Scanning OpenClaw credentials dir: %s", creds_dir)

        # WhatsApp credentials
        wa_dir = creds_dir / "whatsapp"
        if wa_dir.exists():
            for account_dir in wa_dir.iterdir():
                if not account_dir.is_dir():
                    continue
                creds_file = account_dir / "creds.json"
                if creds_file.exists():
                    self._scan_json_file(
                        creds_file, result, show_secrets,
                        context=f"whatsapp:{account_dir.name}",
                    )

        # Scan all JSON files in credentials root (oauth.json, allowlists, etc.)
        for json_file in creds_dir.glob("*.json"):
            self._scan_json_file(json_file, result, show_secrets, context="credentials")

    # --- secrets.json ---

    def _scan_secrets_json(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        path = base / "secrets.json"
        if not path.exists():
            return

        logger.debug("Reading secrets.json: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.warning("Failed to parse %s: %s", path, e, exc_info=True)
            result.errors.append(f"Failed to parse {path}: {e}")
            return

        if isinstance(data, dict):
            self._extract_secrets_recursive(
                data, path, perms, owner, mtime, result, show_secrets,
                context="secrets.json",
            )

    # --- Main config (openclaw.json) ---

    def _scan_main_config(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        path = base / "openclaw.json"
        if not path.exists():
            return

        logger.debug("Reading main config: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        # openclaw.json is JSON5 but stdlib json handles most cases
        try:
            content = path.read_text(encoding="utf-8")
            # Strip JS-style comments for basic JSON5 compat
            lines = []
            for line in content.splitlines():
                stripped = line.lstrip()
                if stripped.startswith("//"):
                    continue
                lines.append(line)
            data = json.loads("\n".join(lines))
        except (json.JSONDecodeError, OSError) as e:
            logger.warning("Failed to parse %s: %s", path, e, exc_info=True)
            result.errors.append(f"Failed to parse {path}: {e}")
            return

        if isinstance(data, dict):
            # Check gateway auth token
            gateway = data.get("gateway", {})
            if isinstance(gateway, dict):
                auth = gateway.get("auth", {})
                if isinstance(auth, dict):
                    token = auth.get("token")
                    if token and isinstance(token, str) and len(token) > 8:
                        notes = ["OpenClaw gateway auth token (inline)"]
                        if mtime:
                            notes.append(f"File last modified: {describe_staleness(mtime)}")
                        result.findings.append(CredentialFinding(
                            tool_name=self.name(),
                            credential_type="gateway_auth_token",
                            storage_type=StorageType.PLAINTEXT_JSON,
                            location=str(path),
                            exists=True,
                            risk_level=assess_risk(StorageType.PLAINTEXT_JSON, path),
                            value_preview=mask_value(token, show_full=show_secrets),
                            raw_value=token if show_secrets else None,
                            file_permissions=perms,
                            file_owner=owner,
                            file_modified=mtime,
                            remediation="Use SecretRef (env:, file:) instead of inline secrets",
                            remediation_hint=hint_manual(
                                "Use SecretRef (env:, file:) instead of inline secrets",
                                suggested_format="env:VAR_NAME",
                            ),
                            notes=notes,
                        ))

            # Check channel configs for inline tokens
            channels = data.get("channels", {})
            if isinstance(channels, dict):
                self._extract_secrets_recursive(
                    channels, path, perms, owner, mtime, result, show_secrets,
                    context="channels",
                )

            # Check agent model configs for API keys
            agents = data.get("agents", {})
            if isinstance(agents, dict):
                self._extract_secrets_recursive(
                    agents, path, perms, owner, mtime, result, show_secrets,
                    context="agents",
                )

    # --- .env file ---

    def _scan_env_file(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        path = base / ".env"
        if not path.exists():
            return

        logger.debug("Reading .env: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)
        storage = StorageType.PLAINTEXT_ENV

        try:
            content = path.read_text(encoding="utf-8")
        except OSError:
            return

        secret_patterns = ["KEY", "TOKEN", "SECRET", "PASSWORD", "AUTH", "CREDENTIAL"]
        for line in content.splitlines():
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" in line:
                key, _, value = line.partition("=")
                key = key.strip()
                value = value.strip().strip("'\"")
                if value and any(p in key.upper() for p in secret_patterns):
                    notes = ["From OpenClaw .env file"]
                    if mtime:
                        notes.append(f"File last modified: {describe_staleness(mtime)}")
                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=f"env_file:{key}",
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(value, show_full=show_secrets),
                        raw_value=value if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation="Use SecretRef (env:, file:) instead of inline secrets",
                        remediation_hint=hint_manual(
                            "Use SecretRef (env:, file:) instead of inline secrets",
                            suggested_format="env:VAR_NAME",
                        ),
                        notes=notes,
                    ))

    # --- Legacy oauth.json ---

    def _scan_legacy_oauth(
        self, base: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        path = base / "credentials" / "oauth.json"
        if not path.exists():
            return

        logger.debug("Reading legacy oauth.json: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.warning("Failed to parse %s: %s", path, e, exc_info=True)
            return

        if isinstance(data, dict):
            self._extract_secrets_recursive(
                data, path, perms, owner, mtime, result, show_secrets,
                context="legacy_oauth",
            )

    # --- Helpers ---

    def _scan_json_file(
        self, path: Path, result: ScanResult, show_secrets: bool, context: str = "",
    ) -> None:
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return

        if isinstance(data, dict):
            self._extract_secrets_recursive(
                data, path, perms, owner, mtime, result, show_secrets,
                context=context,
            )

    def _extract_secrets_recursive(
        self, data: dict, path: Path, perms, owner, mtime,
        result: ScanResult, show_secrets: bool,
        context: str = "", depth: int = 0,
    ) -> None:
        if depth > 10:
            return

        for key, value in data.items():
            if isinstance(value, str) and len(value) > 8:
                if self._is_secret_key(key) or self._looks_like_token(value):
                    # Skip SecretRef values (they reference external sources, not inline secrets)
                    if value.startswith("env:") or value.startswith("file:") or value.startswith("exec:"):
                        notes = [
                            f"Context: {context}" if context else "",
                            "SecretRef (not inline — references external source)",
                        ]
                        if mtime:
                            notes.append(f"File last modified: {describe_staleness(mtime)}")
                        result.findings.append(CredentialFinding(
                            tool_name=self.name(),
                            credential_type=f"secret_ref:{key}",
                            storage_type=StorageType.PLAINTEXT_JSON,
                            location=str(path),
                            exists=True,
                            risk_level=RiskLevel.INFO,
                            value_preview=value,
                            file_modified=mtime,
                            notes=notes,
                        ))
                        continue

                    storage = StorageType.PLAINTEXT_JSON
                    notes = []
                    if context:
                        notes.append(f"Context: {context}")
                    notes.append("Plaintext credential in OpenClaw config")
                    if mtime:
                        notes.append(f"File last modified: {describe_staleness(mtime)}")

                    result.findings.append(CredentialFinding(
                        tool_name=self.name(),
                        credential_type=key,
                        storage_type=storage,
                        location=str(path),
                        exists=True,
                        risk_level=assess_risk(storage, path),
                        value_preview=mask_value(value, show_full=show_secrets),
                        raw_value=value if show_secrets else None,
                        file_permissions=perms,
                        file_owner=owner,
                        file_modified=mtime,
                        remediation="Use SecretRef (env:, file:) instead of inline secrets",
                        remediation_hint=hint_manual(
                            "Use SecretRef (env:, file:) instead of inline secrets",
                            suggested_format="env:VAR_NAME",
                        ),
                        notes=notes,
                    ))

            elif isinstance(value, dict):
                self._extract_secrets_recursive(
                    value, path, perms, owner, mtime, result, show_secrets,
                    context=context, depth=depth + 1,
                )
            elif isinstance(value, list):
                for item in value:
                    if isinstance(item, dict):
                        self._extract_secrets_recursive(
                            item, path, perms, owner, mtime, result, show_secrets,
                            context=context, depth=depth + 1,
                        )

    @staticmethod
    def _is_secret_key(key: str) -> bool:
        key_lower = key.lower()
        secret_keywords = [
            "token", "key", "secret", "password", "passwd", "auth",
            "credential", "cred", "apikey", "api_key", "accesstoken",
            "refreshtoken", "bottoken", "clientsecret",
        ]
        return any(kw in key_lower for kw in secret_keywords)

    @staticmethod
    def _looks_like_token(value: str) -> bool:
        if len(value) < 20:
            return False
        if value.startswith(("/", "http://", "https://", "env:", "file:", "exec:")):
            return False
        alphanumeric_ratio = sum(c.isalnum() or c in "-_." for c in value) / len(value)
        return alphanumeric_ratio > 0.8
